### PR TITLE
Add group members

### PR DESCRIPTION
**Pull request by rosasck**
no test yet

### CODE REVIEW

File: lib/storage/client.js
Comment (carrala) [inline lib/storage/client.js]: I think you forgot to finish out this sentence.

File: __tests__/utils/operations.js
Comment (carrala) [inline __tests__/utils/operations.js:553]: is groupMember able to parse an array of groupMembers correctly? my understanding was that it only sets up one groupMember at a time; however, i haven't run your code yet, so it's very possible that this does actually work.

Comment (rosasck) [reply __tests__/utils/operations.js:553]: Currently attempting to test that, i think you are right tho!

Comment (carrala) [reply __tests__/utils/operations.js:553]: I had a similar issue with listGroups. I think you can do something similar to what you're doing to decode the members in client.js (line 1229).

File: lib/storage/client.js
Comment (galxy25) [inline lib/storage/client.js:1198]: I applaud the effort, but its javascript so better to embrace its untyped nature and not sporadically type check things (e.g. we're not type checking the groupId but we are for group members). Also I think this will fail / doesn't lined up with the documentation as the docs and variable name indicate it can be more than one group member, in which case instanceof == Array and that would cause this method to fail.

File: package.json
Comment (galxy25) [inline package.json:58]: Is this being used?

File: lib/storage/client.js
Comment (galxy25) [inline lib/storage/client.js:1232]: Take a look at what `validateResponseAsJSON` does, I think that because it catches any non 2xx response and throws an exception that you can and should remove the switch statement as you're only going to make it past line 1227 if the response is a 2xx level. (Which obviously I'm opposed to because it means that we've given up control to a function we call unless we want to wrap and catch the exception ourself, but as with everything, tradeoffs and time make pragmatists of us all, so it is what it is and we can work around it)

**Review by galxy25 — APPROVED** [review]
approved pending passing test

File: package.json
Comment (rosasck) [reply package.json:58]: ran npm install and that changed

**Comment by rosasck** [comment]
Reworked to create membership key

**Comment by rosasck** [comment]
right now its functional for a non bulk endpoint! Tests pass. Will work on making it a bulk endpoint next

File: lib/storage/client.js
Comment (willmichael) [inline lib/storage/client.js:1210]: This method will likely fail when done in batch -- The batch call is a POST endpoint living at /v1/client/public and it accepts:
```
type batchPublicRequest struct {
	ClientIDs []uuid.UUID `json:"client_ids"`
}
```

File: types/groupMember.js
Comment (willmichael) [inline types/groupMember.js:8]: Any reason to not init the membership key here?

File: lib/storage/client.js
Comment (willmichael) [inline lib/storage/client.js:1221]: Who is the authorizer in this scenario and why do we need their information here, isn't the caller of this method the one adding new group members and thus the authorizer for these clients?

**Review by willmichael — COMMENTED** [review]
Added some comments to aid in adding the bulk endpoints, its important to note that at the end of the day the endpoint really just accepts an encrypted payload and it cannot do validation of the input, so additional verification is needed perhaps with other SDKS to verify the addition of the group member is validated -- We shouldn't merge this into master as is

**Comment by rosasck** [comment]
Remove group members bulk endpoint is complete!, working on making add group member bulk now

File: lib/storage/client.js
Comment (rosasck) [reply lib/storage/client.js:1221]: we need to know who wrote the membership key for the person adding these people. Since we need their public key in order to unwrap the membership key

File: lib/storage/client.js
Comment (rosasck) [reply lib/storage/client.js:1210]: i couldnt get the array to work so i opted to do this to atleast get some functionality working!, but fixing that now

**Comment by rosasck** [comment]
Add group member bulk endpoint complete  + tests passing

**Comment by rosasck** [comment]
list group member and tests

**Comment by rosasck** [comment]
list shared records with group is complete + test, i did have to do a small work around until i get the PR in storage service merged on lines 
1387-1389 in client.js 
this will be removed once that gets merged

**Comment by rosasck** [comment]
JS SDK Completed Support:
- Add Group Member
- Remove Group Member
- List Group members
- List all Records shared with group

**Comment by galxy25** [comment]
> JS SDK Completed Support:
> 
> * Add Group Member
> * Remove Group Member
> * List Group members
> * List all Records shared with group
> 
> Todo:
> 
> * Share Record with Group

@rosasck Excellent work, and reminder to merge early and often, this may be a good point to get this current pr merged and have a smaller one for faster review for the share record with group task

**Comment by rosasck** [comment]
JS SDK Completed Support:

Add Group Member
Remove Group Member
List Group members
List all Records shared with group
Share Record with Group

File: __tests__/storage.test.js
Comment (galxy25) [inline __tests__/storage.test.js:415]: this is some elegant test code. If you don't have to debug it, javascript can be quite pleasant to read at times 🤔

File: lib/storage/client.js
Comment (galxy25) [inline lib/storage/client.js:1249]: These are magic and crucial values that should be extracted out to named constant, i.e. https://github.com/tozny/e3db-clients-go/blob/master/crypto.go#L22

**Review by galxy25 — APPROVED** [review]
👏🏽

**Comment by rosasck** [comment]
Travis is failing since my storage service is only deployed to dev!

File: lib/crypto/crypto.js
Comment (willmichael) [inline lib/crypto/crypto.js:743]: string

File: lib/storage/client.js
Comment (willmichael) [inline lib/storage/client.js:1242]: do we need this comment?

File: lib/storage/client.js
Comment (willmichael) [inline lib/storage/client.js:1327]: Useful to add the status code to of response to provide some context of the failure, also we're "removing group members" here

File: lib/storage/client.js
Comment (willmichael) [inline lib/storage/client.js:1350]: Is this necessary, aren't you just iterating over a list and pushing contents onto another list?

File: lib/storage/client.js
Comment (willmichael) [inline lib/storage/client.js:1423]: Will this method also create a reader key if it doesn't exist?

File: lib/storage/client.js
Comment (willmichael) [inline lib/storage/client.js:1445]: recordsJson doesn't reflect what's being returned here, also the assignment is unecessary

File: types/groupMember.js
Comment (willmichael) [reply types/groupMember.js:8]: Sorry I forget, did we have a discussion outside of why this shouldn't be part of the init?

File: __tests__/storage.test.js
Comment (willmichael) [inline __tests__/storage.test.js:377]: The results of this call seem to be unused?

File: __tests__/storage.test.js
Comment (willmichael) [inline __tests__/storage.test.js:386]: How are we identifying that records are listed successfully if we haven't written any, unless I'm missing them in this test?

File: __tests__/storage.test.js
Comment (willmichael) [inline __tests__/storage.test.js:414]: The share operation was successful, but how do we know that records are actually shared without listing records?

**Review by willmichael — COMMENTED** [review]
Added some comment on comments, and questions about the tests

File: lib/storage/client.js
Comment (willmichael) [inline lib/storage/client.js:1408]: This methods needs to return the next token, so the caller of this method can paginate through his results

File: lib/storage/client.js
Comment (rosasck) [reply lib/storage/client.js:1423]: nope

**Pull request merged**
